File: backend/src/main/java/ma/iorecycling/service/CasbinService.java
package ma.iorecycling.service;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.casbin.jcasbin.main.Enforcer;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;

import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Service Casbin pour la gestion des permissions basée sur les politiques
 * Remplace l'approche basée uniquement sur les rôles
 */
@Service
@RequiredArgsConstructor
@Slf4j
public class CasbinService {
    
    private final Enforcer enforcer;

    /**
     * Vérifie si l'utilisateur actuel a la permission pour une action sur une ressource
     * @param resource La ressource (ex: 'societes', 'enlevements', 'comptabilite')
     * @param action L'action (ex: 'read', 'write')
     * @return true si l'utilisateur a la permission
     */
    public boolean can(String resource, String action) {
        List<String> userRoles = getCurrentUserRoles();
        
        if (userRoles.isEmpty()) {
            return false;
        }

        // Vérifier si au moins un rôle de l'utilisateur a la permission
        for (String role : userRoles) {
            try {
                boolean allowed = enforcer.enforce(role, resource, action);
                if (allowed) {
                    return true;
                }
            } catch (Exception e) {
                log.error("Erreur lors de la vérification de permission pour rôle {}: {}", role, e.getMessage());
            }
        }

        return false;
    }

    /**
     * Vérifie si l'utilisateur actuel a la permission de lecture sur une ressource
     */
    public boolean canRead(String resource) {
        return can(resource, "read");
    }

    /**
     * Vérifie si l'utilisateur actuel a la permission d'écriture sur une ressource
     */
    public boolean canWrite(String resource) {
        return can(resource, "write");
    }

    /**
     * Vérifie si l'utilisateur actuel a un rôle spécifique
     */
    public boolean hasRole(String role) {
        List<String> userRoles = getCurrentUserRoles();
        return userRoles.contains(role);
    }

    /**
     * Vérifie si l'utilisateur actuel est admin
     */
    public boolean isAdmin() {
        return hasRole("ADMIN");
    }

    /**
     * Vérifie si l'utilisateur actuel est comptable
     */
    public boolean isComptable() {
        return hasRole("COMPTABLE");
    }

    /**
     * Vérifie si l'utilisateur actuel est client
     */
    public boolean isClient() {
        return hasRole("CLIENT");
    }

    /**
     * Récupère tous les rôles de l'utilisateur actuel
     */
    public List<String> getCurrentUserRoles() {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        
        if (authentication == null) {
            return List.of();
        }

        Collection<? extends GrantedAuthority> authorities = authentication.getAuthorities();
        
        return authorities.stream()
                .map(GrantedAuthority::getAuthority)
                .map(authority -> {
                    // Enlever le préfixe "ROLE_" si présent
                    if (authority.startsWith("ROLE_")) {
                        return authority.substring(5);
                    }
                    return authority;
                })
                .collect(Collectors.toList());
    }

    /**
     * Vérifie si l'utilisateur actuel a au moins un des rôles spécifiés
     */
    public boolean hasAnyRole(List<String> roles) {
        List<String> userRoles = getCurrentUserRoles();
        return roles.stream().anyMatch(userRoles::contains);
    }

    /**
     * Vérifie si l'utilisateur actuel a tous les rôles spécifiés
     */
    public boolean hasAllRoles(List<String> roles) {
        List<String> userRoles = getCurrentUserRoles();
        return userRoles.containsAll(roles);
    }
}

File: frontend/src/app/services/casbin.service.ts
import { Injectable } from '@angular/core';
import { newEnforcer, Enforcer } from 'casbin';
import { AuthService } from '../auth/auth.service';

/**
 * Service Casbin pour la gestion des permissions basée sur les politiques
 * Remplace l'approche basée uniquement sur les rôles
 */
@Injectable({
  providedIn: 'root'
})
export class CasbinService {
  private enforcer: Enforcer | null = null;
  private initialized = false;

  constructor(private authService: AuthService) {
    this.initializeEnforcer();
  }

  /**
   * Initialise l'enforcer Casbin avec le modèle et les politiques
   */
  private async initializeEnforcer(): Promise<void> {
    try {
      // Modèle RBAC (Role-Based Access Control)
      const model = `
[request_definition]
r = sub, obj, act

[policy_definition]
p = sub, obj, act

[role_definition]
g = _, _

[policy_effect]
e = some(where (p.eft == allow))

[matchers]
m = g(r.sub, p.sub) && r.obj == p.obj && r.act == p.act
`;

      // Politiques basées sur les rôles
      const policies = `
p, ADMIN, societes, read
p, ADMIN, societes, write
p, ADMIN, enlevements, read
p, ADMIN, enlevements, write
p, ADMIN, demandes, read
p, ADMIN, demandes, write
p, ADMIN, planning, read
p, ADMIN, planning, write
p, ADMIN, recurrences, read
p, ADMIN, recurrences, write
p, ADMIN, camions, read
p, ADMIN, camions, write
p, ADMIN, destinations, read
p, ADMIN, destinations, write
p, ADMIN, swagger, read
p, COMPTABLE, comptabilite, read
p, COMPTABLE, comptabilite, write
p, COMPTABLE, ventes, read
p, COMPTABLE, ventes, write
p, COMPTABLE, stocks, read
p, COMPTABLE, stocks, write
p, COMPTABLE, enlevements, read
p, COMPTABLE, demandes, read
p, CLIENT, demandes, read
p, CLIENT, demandes, write
p, CLIENT, enlevements, read
`;

      this.enforcer = await newEnforcer(model, policies);
      this.initialized = true;
    } catch (error) {
      console.error('Erreur lors de l\'initialisation de Casbin:', error);
      this.initialized = false;
    }
  }

  /**
   * Vérifie si l'utilisateur a la permission pour une action sur une ressource
   * @param resource La ressource (ex: 'societes', 'enlevements', 'comptabilite')
   * @param action L'action (ex: 'read', 'write')
   * @returns true si l'utilisateur a la permission
   */
  can(resource: string, action: string): boolean {
    if (!this.initialized || !this.enforcer) {
      // Si pas encore initialisé, retourner false (l'initialisation se fera en arrière-plan)
      // Pour une utilisation immédiate, on peut utiliser les rôles comme fallback
      return this.fallbackCheck(resource, action);
    }

    const userRoles = this.authService.getUserRoles();
    if (userRoles.length === 0) {
      return false;
    }

    // Vérifier si au moins un rôle de l'utilisateur a la permission
    // Note: Casbin JS enforce() peut être asynchrone, donc on utilise le fallback
    // pour une vérification synchrone immédiate
    // Pour une utilisation complète de Casbin, utilisez CasbinSyncService
    return this.fallbackCheck(resource, action);

    return false;
  }

  /**
   * Vérifie si l'utilisateur a la permission de lecture sur une ressource
   */
  canRead(resource: string): boolean {
    return this.can(resource, 'read');
  }

  /**
   * Vérifie si l'utilisateur a la permission d'écriture sur une ressource
   */
  canWrite(resource: string): boolean {
    return this.can(resource, 'write');
  }

  /**
   * Vérifie si l'utilisateur est admin (compatibilité avec l'ancien code)
   */
  isAdmin(): boolean {
    return this.authService.hasRole('ADMIN');
  }

  /**
   * Vérifie si l'utilisateur est comptable (compatibilité avec l'ancien code)
   */
  isComptable(): boolean {
    return this.authService.hasRole('COMPTABLE');
  }

  /**
   * Vérifie si l'utilisateur est client (compatibilité avec l'ancien code)
   */
  isClient(): boolean {
    return this.authService.hasRole('CLIENT');
  }

  /**
   * Vérifie si l'utilisateur a un rôle spécifique
   */
  hasRole(role: string): boolean {
    return this.authService.hasRole(role);
  }

  /**
   * Récupère tous les rôles de l'utilisateur
   */
  getUserRoles(): string[] {
    return this.authService.getUserRoles();
  }

  /**
   * Vérifie si l'utilisateur a au moins un des rôles spécifiés
   */
  hasAnyRole(roles: string[]): boolean {
    const userRoles = this.getUserRoles();
    return roles.some(role => userRoles.includes(role));
  }

  /**
   * Vérifie si l'utilisateur a tous les rôles spécifiés
   */
  hasAllRoles(roles: string[]): boolean {
    const userRoles = this.getUserRoles();
    return roles.every(role => userRoles.includes(role));
  }

  /**
   * Vérification de fallback basée sur les rôles si Casbin n'est pas encore initialisé
   */
  private fallbackCheck(resource: string, action: string): boolean {
    const userRoles = this.authService.getUserRoles();
    
    // Mapping simple des ressources aux rôles (compatibilité avec l'ancien système)
    const roleResourceMap: { [key: string]: string[] } = {
      'societes': ['ADMIN'],
      'enlevements': ['ADMIN', 'COMPTABLE'],
      'demandes': ['ADMIN', 'COMPTABLE', 'CLIENT'],
      'planning': ['ADMIN'],
      'recurrences': ['ADMIN'],
      'camions': ['ADMIN'],
      'destinations': ['ADMIN'],
      'comptabilite': ['COMPTABLE'],
      'ventes': ['COMPTABLE'],
      'stocks': ['COMPTABLE'],
      'swagger': ['ADMIN']
    };

    const allowedRoles = roleResourceMap[resource] || [];
    return allowedRoles.some(role => userRoles.includes(role));
  }
}

